src/control/server/mgmt_svc_test.go

```diff
@@ -1933,6 +1933,166 @@ func TestMgmtSvc_KillRanks(t *testing.T) {
+				},
+			},
+		},
+		"dRPC resp fails": {
+			req:     &mgmtpb.RanksReq{},
+			drpcRet: errors.New("call failed"),
+			drpcResps: []proto.Message{
+				&mgmtpb.DaosResp{Status: 0},
+				&mgmtpb.DaosResp{Status: 0},
+			},
+			expResp: &mgmtpb.RanksResp{
+				Results: []*mgmtpb.RanksResp_RankResult{
+					{Rank: 1, Action: "ping", State: 5, Errored: true},
+					{Rank: 2, Action: "ping", State: 5, Errored: true},
+				},
+			},
+		},
+		"dRPC resp junk": {
+			req:      &mgmtpb.RanksReq{},
+			junkResp: true,
+			expResp: &mgmtpb.RanksResp{
+				Results: []*mgmtpb.RanksResp_RankResult{
+					{Rank: 1, Action: "ping", State: 5, Errored: true},
+					{Rank: 2, Action: "ping", State: 5, Errored: true},
+				},
+			},
+		},
+		"unsuccessful call": {
+			req: &mgmtpb.RanksReq{},
+			drpcResps: []proto.Message{
+				&mgmtpb.DaosResp{Status: -1},
+				&mgmtpb.DaosResp{Status: -1},
+			},
+			expResp: &mgmtpb.RanksResp{
+				Results: []*mgmtpb.RanksResp_RankResult{
+					{Rank: 1, Action: "ping", State: 5, Errored: true},
+					{Rank: 2, Action: "ping", State: 5, Errored: true},
+				},
+			},
+		},
+		"successful call": {
+			req: &mgmtpb.RanksReq{},
+			drpcResps: []proto.Message{
+				&mgmtpb.DaosResp{Status: 0},
+				&mgmtpb.DaosResp{Status: 0},
+			},
+			expResp: &mgmtpb.RanksResp{
+				Results: []*mgmtpb.RanksResp_RankResult{
+					{Rank: 1, Action: "ping", State: 1},
+					{Rank: 2, Action: "ping", State: 1},
+				},
+			},
+		},
+		"ping timeout": {
+			req:           &mgmtpb.RanksReq{},
+			responseDelay: 200 * time.Millisecond,
+			drpcResps: []proto.Message{
+				&mgmtpb.DaosResp{Status: 0},
+				&mgmtpb.DaosResp{Status: 0},
+			},
+			expResp: &mgmtpb.RanksResp{
+				Results: []*mgmtpb.RanksResp_RankResult{
+					{Rank: 1, Action: "ping", State: 6, Errored: true},
+					{Rank: 2, Action: "ping", State: 6, Errored: true},
+				},
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			log, buf := logging.NewTestLogger(t.Name())
+			defer common.ShowBufferOnFailure(t, buf)
+
+			ioserverCount := maxIoServers
+			svc := newTestMgmtSvcMulti(log, ioserverCount, tc.setupAP)
+			for i, srv := range svc.harness.instances {
+				if tc.missingSB {
+					srv._superblock = nil
+					continue
+				}
+				if tc.instancesStopped { // real runner reports not started
+					srv.runner = ioserver.NewRunner(log,
+						ioserver.NewConfig())
+				}
+
+				srv._superblock.Rank = new(ioserver.Rank)
+				*srv._superblock.Rank = ioserver.Rank(i + 1)
+
+				cfg := new(mockDrpcClientConfig)
+				if tc.drpcRet != nil {
+					cfg.setSendMsgResponse(drpc.Status_FAILURE, nil, nil)
+				} else if tc.junkResp {
+					cfg.setSendMsgResponse(drpc.Status_SUCCESS, makeBadBytes(42), nil)
+				} else if len(tc.drpcResps) > i {
+					rb, _ := proto.Marshal(tc.drpcResps[i])
+					cfg.setSendMsgResponse(drpc.Status_SUCCESS, rb, tc.expErr)
+
+					if tc.responseDelay != time.Duration(0) {
+						cfg.setResponseDelay(tc.responseDelay)
+					}
```

Comment:
I'm a little uncomfortable with adding realtime delays to unit tests. This one is less than a second, so not that big a deal, but this is the kind of thing one hopes to mock, generally. Doesn't look like that's feasible to do here without some extra work, but might be something to think about as we move forward. If we have to test anything with timeouts on the order of seconds or minutes in the future, the real delays will get ugly.